### PR TITLE
UI for challenge workflow

### PR DESCRIPTION
**Pull request by lankastersky**

### CODE REVIEW

File: zen/utils/UIView+Utils.swift [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import Foundation
+import UIKit
+
+extension UIView {
+    class func fromNib<T: UIView>() -> T {
+        guard let views = Bundle.main.loadNibNamed(String(describing: T.self), owner: nil, options: nil) else {
+            assertionFailure("Failed to instantiate view")
+            return T()
+        }
+        guard let view = views[0] as? T else {
```
Comment (NikAshanin) [inline zen/utils/UIView+Utils.swift]: views.first

File: zen/views/challenge_footer_views/FinishedChallengeFooterView.swift [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import UIKit
+import Cosmos
+
+final class FinishedChallengeFooterView: ChallengeFooterView {
+
+    @IBOutlet private weak var ratingView: CosmosView!
+
+    @IBOutlet weak var commentsLabel: UILabel!
```
Comment (NikAshanin) [inline zen/views/challenge_footer_views/FinishedChallengeFooterView.swift:8]: private

**Review by NikAshanin — CHANGES_REQUESTED** [review]
https://habr.com/company/touchinstinct/blog/327356/ посмотри здесь, как можно сделать локализацию элегантнее без magic strings в коде

File: zen/views/challenge_footer_views/FinishedChallengeFooterView.swift (hunk 2 above)
Comment (lankastersky) [reply zen/views/challenge_footer_views/FinishedChallengeFooterView.swift:8]: It's initialized in ChallengeViewController

File: zen/utils/UIView+Utils.swift (hunk 1 above)
Comment (lankastersky) [reply zen/utils/UIView+Utils.swift]: done

File: zen/services/FirebaseService.swift [hunk 3]
```diff
@@ -47,33 +56,73 @@ final class FirebaseService {
...
             if let error = error {
-                callback(nil, error)
+                callback(error)
             } else if let url = url {
                 do {
                     let unzippedData = try FirebaseService
                         .unzip(url, "\(FirebaseService.challengesFileName).json")
                     challenges = try FirebaseService.parseChallenges(data: unzippedData)
-                    callback(challenges, nil)
+                    self.challengesService.storeChallenges(challenges)
```
Comment (NikAshanin) [inline zen/services/FirebaseService.swift:82]: weak

File: zen/services/FirebaseService.swift [hunk 4]
```diff
@@ -47,33 +56,73 @@ final class FirebaseService {
...
+                    self.storeLastLoadTime()
```
Comment (NikAshanin) [inline zen/services/FirebaseService.swift:83]: weak

File: zen/views/ChallengeViewController.swift [hunk 5]
```diff
@@ -5,118 +5,210 @@ final class ChallengeViewController: UIViewController {
...
-        navigationItem.title = "current_challenge_screen_title".localized
+        if challenge == nil {
+            navigationItem.title = "current_challenge_screen_title".localized
+        } else {
+            navigationItem.title = "challenge_screen_title".localized
+        }
+
+        let tapGesture =
+            UITapGestureRecognizer(target: self, action: #selector(self.dismissKeyboard (_:)))
+        self.view.addGestureRecognizer(tapGesture)
```
Comment (NikAshanin) [inline zen/views/ChallengeViewController.swift]: self redundant

File: zen/views/ChallengeViewController.swift [hunk 6]
```diff
@@ -5,118 +5,210 @@ final class ChallengeViewController: UIViewController {
...
 
-        navigationItem.title = "current_challenge_screen_title".localized
+        if challenge == nil {
+            navigationItem.title = "current_challenge_screen_title".localized
+        } else {
+            navigationItem.title = "challenge_screen_title".localized
+        }
+
+        let tapGesture =
+            UITapGestureRecognizer(target: self, action: #selector(self.dismissKeyboard (_:)))
```
Comment (NikAshanin) [inline zen/views/ChallengeViewController.swift]: self redundant

File: zen/views/ChallengeViewController.swift [hunk 7]
```diff
@@ -5,118 +5,210 @@ final class ChallengeViewController: UIViewController {
...
-            challengeButton.setTitle(
-                "current_challenge_screen_button_finished".localized,
-                for: .normal)
-            challengeButton.isEnabled = false
-        default:
-            assertionFailure("Bad challenge status when refreshing challenge button")
-        }
+        let keyboardSize = keyboardInfo.cgRectValue.size
+        let topBarHeight = UIApplication.shared.statusBarFrame.size.height +
+            (self.navigationController?.navigationBar.frame.height ?? 0.0)
```
Comment (NikAshanin) [inline zen/views/ChallengeViewController.swift]: self redundant

**Review by NikAshanin — CHANGES_REQUESTED** [review]

File: zen/services/FirebaseService.swift (hunk 3 above)
Comment (lankastersky) [reply zen/services/FirebaseService.swift:82]: if i make it weak, it becomes nil and doesn't work

File: zen/services/FirebaseService.swift (hunk 4 above)
Comment (lankastersky) [reply zen/services/FirebaseService.swift:83]: if i make it weak, it becomes nil and doesn't work

File: zen/views/ChallengeViewController.swift (hunk 5 above)
Comment (lankastersky) [reply zen/views/ChallengeViewController.swift]: done

File: zen/views/ChallengeViewController.swift (hunk 6 above)
Comment (lankastersky) [reply zen/views/ChallengeViewController.swift]: done

File: zen/views/ChallengeViewController.swift (hunk 7 above)
Comment (lankastersky) [reply zen/views/ChallengeViewController.swift]: done

**Pull request merged**